pinot-common/src/main/java/org/apache/pinot/common/config/provider/UserCache.java

```diff
@@ -0,0 +1,176 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.common.config.provider;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import org.I0Itec.zkclient.IZkChildListener;
+import org.I0Itec.zkclient.IZkDataListener;
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.helix.AccessOption;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.store.zk.ZkHelixPropertyStore;
+import org.apache.pinot.common.utils.config.UserConfigUtils;
+import org.apache.pinot.spi.config.user.ComponentType;
+import org.apache.pinot.spi.config.user.UserConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class UserCache {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(UserCache.class);
+    private static final String USER_CONFIG_PARENT_PATH = "/CONFIGS/USER";
+    private static final String USER_CONFIG_PATH_PREFIX = "/CONFIGS/USER/";
+
+    private final ZkHelixPropertyStore<ZNRecord> _propertyStore;
+
+    private final UserConfigChangeListener _userConfigChangeListener = new UserConfigChangeListener();
+
+    private final Map<String, UserConfig> _userConfigMap = new ConcurrentHashMap<>();
+    private final Map<String, UserConfig> _userControllerConfigMap = new ConcurrentHashMap<>();
+    private final Map<String, UserConfig> _userBrokerConfigMap = new ConcurrentHashMap<>();
+    private final Map<String, UserConfig> _userServerConfigMap = new ConcurrentHashMap<>();
```

Comment:
if you intend to configure all components dynamically, you'll need to track `minion` as well.

@xiangfu0 should we track dedicated permissions for the upcoming pinot proxy as well?

Reply:
@apucher I think your suggestion is more reasonable，but now Pinot haven't added access control to minion and proxy. We would get nothing information about minion and proxy. Maybe I didn't understand your point. I look forward to your reply. Thx

Reply:
it's fine. we'll extend this whenever we need any properties for minion or proxy